### PR TITLE
Diff.py: Add coala's own unified_diff method

### PR DESCRIPTION
**Pull request by aptrishu**
### Check list

- [x] I have rebased properly.
      [Please see this for a tutorial.](http://coala.readthedocs.io/en/latest/Developers/Git_Basics.html#rebasing)
- [x] I have gone through the
      [commit guidelines](http://coala.readthedocs.io/en/latest/Developers/Writing_Good_Commits.html)
      and I've followed them:
 - [x] I have followed the
       [guidelines for the commit shortlog](http://docs.coala.io/en/latest/Developers/Writing_Good_Commits.html#shortlog).
 - [x] I have followed the
       [guidelines for the commit body](http://docs.coala.io/en/latest/Developers/Writing_Good_Commits.html#commit-body).
 - [x] I have included the issue URL with the  `Fixes` keyword if the
       commit fixes a *real bug* and the `Closes` keyword if it adds a
       feature/enhancement.
       [See more here.](http://docs.coala.io/en/latest/Developers/Writing_Good_Commits.html#issue-reference)
- [x] I have [run all the tests](http://coala.readthedocs.io/en/latest/Developers/Executing_Tests.html)
      and they all pass. If any CI (below) is not green, please fix them. If
      GitMate issues appear you will have to amend your commits, pushing new
      commits on top is not sufficient!

### Reviewers

It adds a new coala_unified_diff method which adds up a functionality
to check for a newline at EOF.
Earlier coala was using python difflib's unified-diff method.

Fixes https://github.com/coala/coala/issues/2006

### CODE REVIEW

File: coalib/results/Diff.py [hunk 1]
```diff
@@ -197,11 +198,72 @@ def unified_diff(self):
...
+        The unidiff format normally has a header for filenames and modification
+        times.  Any or all of these may be specified using strings for
+        'fromfile', 'tofile', 'fromfiledate', and 'tofiledate'.
+        The modification times are normally expressed in the ISO 8601 format.
+
+        If a or b misses a newline at end of file, an additional newline is
+        generated at end of last line, and an extra warning line "\ No newline
+        at end of file" is generated if warn_no_newline_at_end is True, to
+        make the patch program happy.
+        """
```
Comment (sils) [inline coalib/results/Diff.py]: Nononononono! You didn't copy this from the stdlib, did you? That's not only code duplication but also you're not allowed to do that unless you give them credit and make sure licenses are compatible (which I think they are but still)

**Comment by sils** [comment]
tests fail because of coverage, rebase might solve this but probably you didn't cover the new code with enough testing

File: coalib/results/Diff.py (hunk 1 above)
Comment (aptrishu) [inline coalib/results/Diff.py]: yes, I only had to change a small part of unified_diff, why would I write the whole method ?
what should I do ?

Comment (sils) [inline coalib/results/Diff.py]: I'm not fully sure, maybe you can postprocess their output?

Comment (aptrishu) [inline coalib/results/Diff.py]: I gave them credit. :+1:

**Comment by tulphoon** [comment]
This looks nice. Do a rebase though, as your code is out of date, not sure about the licensing and stuff about usage of code from python's difflib.

**Comment by aptrishu** [comment]
We can use that code, actually I have seen other orgs have also used it. I have given them the credit.

**Pull request closed without merging**